### PR TITLE
Deprecate window global

### PR DESCRIPTION
**Pull request by NullVoxPopuli**

### CODE REVIEW

File: packages/ember/index.js [hunk 1]
```diff
@@ -171,7 +171,7 @@ import {
 const Ember = (typeof context.imports.Ember === 'object' && context.imports.Ember) || {};
 
 Ember.isNamespace = true;
-Ember.toString = function () {
+Ember.toString = function() {
```
Comment (rwjblue) [inline packages/ember/index.js:174]: you'll want to roll back these unrelated prettier changes (also seems like maybe your editor was applying a different prettier version than our linting uses)

File: packages/ember/index.js [hunk 2]
```diff
@@ -457,7 +457,7 @@ Ember.isEqual = isEqual;
 Ember.inject = function inject() {
   assert(
     `Injected properties must be created through helpers, see '${Object.keys(inject)
-      .map((k) => `'inject.${k}'`)
+      .map(k => `'inject.${k}'`)
```
Comment (rwjblue) [inline packages/ember/index.js:460]: 🥒

File: packages/ember/index.js [hunk 3]
```diff
@@ -647,11 +647,11 @@ Ember.HTMLBars = {
 };
 
 if (ENV.EXTEND_PROTOTYPES.String) {
-  String.prototype.htmlSafe = function () {
+  String.prototype.htmlSafe = function() {
```
Comment (rwjblue) [inline packages/ember/index.js:650]: 🍉

File: packages/ember/index.js [hunk 4]
```diff
@@ -647,11 +647,11 @@ Ember.HTMLBars = {
...
     return htmlSafe(this);
   };
 }
-const deprecateImportFromString = function (
+const deprecateImportFromString = function(
```
Comment (rwjblue) [inline packages/ember/index.js:654]: 🍅

File: packages/ember/index.js [hunk 5]
```diff
@@ -791,6 +791,36 @@ if (IS_NODE) {
...
+        writable: false,
+        get() {
+          return Ember;
+        },
+      });
+
+      return Ember;
+    },
+  });
+}
```
Comment (rwjblue) [inline packages/ember/index.js:822]: I think this needs to be just above where `context.exports.Ember = context.exports.Em = Ember;` is.

`context.exports` _is_ `window` IIRC

**Review by rwjblue — CHANGES_REQUESTED** [review]
I think this will also need a couple of tests

File: packages/ember/index.js (hunk 1 above)
Comment (NullVoxPopuli) [reply packages/ember/index.js:174]: for sure, idk what happened here (very WIP PR) :p

File: packages/ember/index.js (hunk 5 above)
Comment (NullVoxPopuli) [reply packages/ember/index.js:822]: oh! this is good to know. thanks!

**Comment by pzuraq** [comment]
So had a quick convo with @rwjblue about this, and we realized we're going to need to make some additional changes. Currently, `ember-cli-babel` still transpiles `import * from '@ember/*'` to use the Ember global, and it does _not_ add `import Ember from 'ember';` at the top. In addition, we don't actually _define_ an `ember` module - that is also transpiled via `ember-cli-babel`.

So, in this package we need to add the Ember package so you can actually import it. In the babel plugin that `ember-cli-babel` adds, we need to update it to use `import Ember from 'ember';` to the top. It should also only do that IF the Ember version is high enough, so probably 3.26.0-beta.1 assuming we get this PR in before then.

**Comment by NullVoxPopuli** [comment]
Next Steps?:
- [ ] update ember-cli-babel
- [ ] add deprecation guide
- [ ] merge this PR?

**Review by rwjblue — CHANGES_REQUESTED** [review]
In order to land we need to:

* Update https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill 
  * Add a new configuration option (`useEmberModule` maybe?) to opt in to emitting the `import Ember from 'ember';` bit or doing the (currently default) "just use `window.Ember`" thing
* Update ember-cli-babel
  * to use that new version of babel-plugin-ember-modules-api-polyfill
  * to pass `true` or `false` to that new option based on the running version of `ember-source`
* Update this PR to actually emit a `ember` module (current Ember apps actually have no `ember` module at runtime, we have to fix that)

**Comment by btecu** [comment]
Aren't there some important parts that have not been exposed via modules yet? How would that work with this?
For example `Ember.onError`, is used for error trackers such as Bugsnag and Sentry.

**Comment by pzuraq** [comment]
@btecu as noted in the RFC, you will still be able to `import Ember from 'ember';`. This is just about accessing the object without importing it.

**Comment by NullVoxPopuli** [comment]
babel plugin: https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/pull/175

**Comment by pzuraq** [comment]
@NullVoxPopuli can you rebase this PR? Now that #19390 has landed, everything should be much easier to deprecate since the global is now setup lazily (you can do it in `@ember/-internals/bootstrap`). Also we need to add a test that verifies the deprecation runs when we use the global value

**Comment by rwjblue** [comment]
https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/releases/tag/v3.3.0 includes new `useEmberModule` option. Next up: update ember-cli-babel.

**Comment by simonihmig** [comment]
This has been superseded by https://github.com/emberjs/ember.js/pull/19457, so can be closed, right?
